### PR TITLE
Cassandra TLS

### PR DESCRIPTION
**Pull request by conorbros**

### CODE REVIEW

**Comment by conorbros** [comment]
Running into a strange issue. I can use Shotover with a TLS Cassandra source fine with `cqlsh` but `cassandra_cpp` does not work. I think it's using the wrong version of TLS (TLS v1.1?).

**Comment by conorbros** [comment]
This PR is now ready for review. 

### Summary of the changes: 
- As discussed on slack, changed the benchmark runner to `ubuntu-18.04` and used the ubuntu install script so we can use `cassandra_cpp` in Cassandra TLS benchmarks, coming in a followup PR.
-  Added SSL keys/certs and the keystore and truststore formats that Cassandra uses. 
- The `cassandra-passthrough-tls` image is `bitnami/cassandra` instead of `library/cassandra`, this is because they have added the ability to enable client encryption through environment variables. I opened a PR to add the other environment variables that we need, so we could switch all our cassandra images to this and get rid of the modified `docker-entrypoint.sh` but I will leave that to a followup PR. 
- Changed the `SslAcceptor` to use `mozilla_intermediate` instead of `mozilla_modern_v5` because `cassandra_cpp` has been limited to use TLS 1.2 or below, as can be seen [here](https://github.com/datastax/cpp-driver/blob/90df2c9ca1aa184a746445698533c71f7f34a2e1/src/ssl/ssl_openssl_impl.cpp#L555). It seems the intermediate [configuration suits well as a general purpose server](https://wiki.mozilla.org/Security/Server_Side_TLS).


### Moving forward
- Add TLS benchmarks
- Move our all our Cassandra images over to the bitnami one.

File: shotover-proxy/tests/cassandra_int_tests/basic_driver_tests.rs
Comment (rukai) [inline shotover-proxy/tests/cassandra_int_tests/basic_driver_tests.rs:948]: println

File: shotover-proxy/tests/helpers/mod.rs
Comment (rukai) [inline shotover-proxy/tests/helpers/mod.rs:145]: Shouldnt this go next to the `fn cassandra_connection` ?

Comment (conorbros) [reply shotover-proxy/tests/helpers/mod.rs:145]: I've moved `fn cassandra_connection` to `ShotoverManger` in #512, so once that is merged they will be in the same place.

Comment (rukai) [reply shotover-proxy/tests/helpers/mod.rs:145]: ah, alright, yeah I think its fine for both of them to go in ShotoverManager.
While its true that we could use cassandra_connection to connect directly to cassandra instead of through shotover, there is no case where we would want to do that without also having a shotover instance running anyway.

File: shotover-proxy/tests/cassandra_int_tests/basic_driver_tests.rs
Comment (rukai) [inline shotover-proxy/tests/cassandra_int_tests/basic_driver_tests.rs:1337]: Can we run a simple test case directly against cassandra to make sure that cassandra_connection_tls can connect to both cassandra and shotover?
This would protect against a possible case where both shotover and cassandra_connection_tls were doing something slightly wrong compared to actual cassandra.
I think `table::test` would be fast enough for this use case.

File: shotover-proxy/tests/helpers/mod.rs
Comment (rukai) [inline shotover-proxy/tests/helpers/mod.rs:162]: why are we introducing username and password as well?
Is it tied to TLS somehow?

Comment (conorbros) [reply shotover-proxy/tests/helpers/mod.rs:162]: The docker-image has it enabled by default, and I thought to leave it on so we can test that as well.

File: shotover-proxy/tests/cassandra_int_tests/basic_driver_tests.rs
Comment (conorbros) [reply shotover-proxy/tests/cassandra_int_tests/basic_driver_tests.rs:1353]: Decided to just run a simple query against the `system.local` table, rather than implement teardown code for the `table::test` suite.

Comment (conorbros) [reply shotover-proxy/tests/cassandra_int_tests/basic_driver_tests.rs:1353]: Decided to just run a simple query against the `system.local` table, rather than implement teardown code for the `table::test` suite.

File: shotover-proxy/tests/helpers/mod.rs
Comment (rukai) [reply shotover-proxy/tests/helpers/mod.rs:163]: Hmmm alright, sounds reasonable

File: shotover-proxy/tests/cassandra_int_tests/basic_driver_tests.rs
Comment (rukai) [reply shotover-proxy/tests/cassandra_int_tests/basic_driver_tests.rs:1353]: nice!

File: shotover-proxy/tests/helpers/mod.rs
Comment (rukai) [inline shotover-proxy/tests/helpers/mod.rs:156]: What is this for?

**Review by benbromhead — COMMENTED** [review]
We also probably need to have a `connect` method on `TlsConnector` that will verify the hostname (I know we dont do it at the moment as redis never did it that way)

File: shotover-proxy/src/tls.rs
Comment (benbromhead) [reply shotover-proxy/src/tls.rs:29]: lol thanks

**Comment by rukai** [comment]
Are all the 6 key files used?

And how are they generated?
Might be worth checking in a shell script in the keys directory for future reference.
I think the reason I didnt do this for redis is there was already a shell script `gen-test-certs` in the redis repo for generating keys.

File: shotover-proxy/src/tls.rs
Comment (rukai) [inline shotover-proxy/src/tls.rs:73]: Can we make setting this to false scarier?
Maybe a separate scary sounding method name?
Maybe a seperate constructor for TLSAcceptor?
Or maybe replace the bool with an enum with a scary sounding variant?

As is, its quite easy to accidentally drop an important security measure by copy-pasting the connect call from redis which doesnt give any context to the false value it uses.

**Comment by conorbros** [comment]
Yeah, I'll remove `*.jks`.

And yes, the process is a little complicated. I'll put together a script for it.

File: shotover-proxy/src/transforms/cassandra/connection.rs
Comment (rukai) [inline shotover-proxy/src/transforms/cassandra/connection.rs:32]: This can no longer be None so we should remove the Option.

File: shotover-proxy/src/transforms/cassandra/connection.rs
Comment (rukai) [inline shotover-proxy/src/transforms/cassandra/connection.rs:41]: we can do `&host` here instead of `host.clone()`

**Review by rukai — CHANGES_REQUESTED** [review]
Mostly reviewed at this point.
Once you've addressed these up i'll go over it again.

File: shotover-proxy/src/tls.rs
Comment (rukai) [inline shotover-proxy/src/tls.rs:86]: this into_ssl("localhost") looks suspicious.
https://docs.rs/openssl/latest/openssl/ssl/struct.ConnectConfiguration.html#method.into_ssl
From reading the docs, it seems this might break in a real non localhost scenario?

It wouldnt have been a problem on redis because hostname verification is disabled there.

If we find an issue and it needs a complex solution/investigation I would be happy to land what we have so far as it at least works in a subset of scenarios and get non-localhost connections working in a follow up PR.

File: shotover-proxy/src/transforms/cassandra/connection.rs
Comment (rukai) [inline shotover-proxy/src/transforms/cassandra/connection.rs:47]: Should we return Err instead of unwrap here?
Currently we never return Err in this function despite being Result<()>

Lets make the redis equivalent call do the same as well.

File: shotover-proxy/src/transforms/util/cluster_connection_pool.rs
Comment (rukai) [inline shotover-proxy/src/transforms/util/cluster_connection_pool.rs:181]: Its concerning that we are making this call in a supposedly generic `util`.
Should we raise an issue to make this cluster logic redis specific?

File: shotover-proxy/examples/cassandra-passthrough-tls/certs/gen_certs.sh
Comment (rukai) [inline shotover-proxy/examples/cassandra-passthrough-tls/certs/gen_certs.sh:1]: best to add bash or sh crunchbang

File: shotover-proxy/examples/cassandra-passthrough-tls/gen_certs.sh
Comment (rukai) [inline shotover-proxy/examples/cassandra-passthrough-tls/gen_certs.sh:1]: Accidental duplicate?

File: shotover-proxy/src/transforms/util/cluster_connection_pool.rs
Comment (conorbros) [reply shotover-proxy/src/transforms/util/cluster_connection_pool.rs:181]: I don't think so, the method is pretty explicit that it does not verify the hostname. I'd say it's as clear as the `verify_hostname` method itself on `openssl::ssl::ConnectConfiguration`.

Comment (rukai) [reply shotover-proxy/src/transforms/util/cluster_connection_pool.rs:181]: right but you would have to check inside the cluster_connection_pool implementation to know that.
We can just leave at as is though, I dont mind.

Comment (conorbros) [reply shotover-proxy/src/transforms/util/cluster_connection_pool.rs:181]: Oh true, I misinterpreted. Thought you were saying the `TlsConnector` was a generic util.

Comment (conorbros) [reply shotover-proxy/src/transforms/util/cluster_connection_pool.rs:181]: When we look at Cassandra cluster support we can revisit that. We will find out whether we can have a generic cluster connection pool or if this needs to be a Redis specific one.

**Review by rukai — APPROVED** [review]
Nice, LGTM!

Lets investigate the unresolved comment, before merging.
But im happy to merge no matter the result and just raise an issue to resolve the problem if there is one.

File: shotover-proxy/src/tls.rs
Comment (conorbros) [reply shotover-proxy/src/tls.rs:86]: image

I'm not sure if that is the right path. Maybe we need to providing SSL connections with a hostname to connect to instead of an address.

Comment (rukai) [reply shotover-proxy/src/tls.rs:86]: were you able to reproduce a problem by connecting to an instance that isnt on local host?

Comment (conorbros) [reply shotover-proxy/src/tls.rs:86]: Yes, trying to connect to with hostname "notlocalhost" creates a problem.

Comment (conorbros) [reply shotover-proxy/src/tls.rs:86]: I'll open an issue regarding TLS configuration. I think we will want an option for validating the hostname as well.

**Comment by benbromhead** [comment]
Sorry one last final question. Given we have the script to generate certs, I'm now hesitant to checkin certs to the repo (even self signed / self generated certs for testing) as they always have a habbit of making their way to live servers in the wild as folks grab them to "just make something work".

Should we just gen the certs as part of the build / CI process and not checkin any certs

**Comment by rukai** [comment]
Oh I hadnt heard of incidents like that, but if thats a thing then generating the certs at runtime definitely makes sense to me.
We will want to do the same for redis as well, so we could either:
*  check in as is and raise an issue to fix both redis and cassandra.
*  fix cassandra now and raise an issue for just redis

I dont mind either way

**Review by benbromhead — APPROVED** [review]
lgtm

**Pull request merged**
